File: packages/validations/src/ats.ts
import { z } from "zod";

const optionalTrimmedString = (maxLength: number) =>
  z
    .string()
    .trim()
    .max(maxLength)
    .optional()
    .transform((value) => value ?? "");

export const atsAnalyzeSchema = z.object({
  resumeId: z.string().uuid(),
  jobTitle: optionalTrimmedString(200),
  companyName: optionalTrimmedString(200),
  jobDescription: z.string().trim().min(20).max(100_000),
  targetKeywords: z.array(z.string().trim().min(1).max(100)).max(80).default([]),
});

export const atsScoreBreakdownSchema = z.object({
  keywords: z.number().int().min(0).max(100),
  semantic: z.number().int().min(0).max(100),
  skills: z.number().int().min(0).max(100),
  experience: z.number().int().min(0).max(100),
  impact: z.number().int().min(0).max(100),
  formatting: z.number().int().min(0).max(100),
});

export const atsSuggestionSchema = z.object({
  priority: z.enum(["high", "medium", "low"]),
  title: z.string(),
  detail: z.string(),
  example: z.string().nullable().optional(),
});

export const atsJobProfileSchema = z.object({
  title: z.string(),
  seniority: z.string(),
  yearsRequired: z.number().int().min(0),
  requiredKeywords: z.array(z.string()),
  preferredKeywords: z.array(z.string()),
  responsibilities: z.array(z.string()),
  certifications: z.array(z.string()),
});

export const atsKeywordEvidenceSchema = z.object({
  keyword: z.string(),
  status: z.enum(["exact", "semantic", "missing"]),
  strength: z.number().int().min(0).max(100),
  sourceSection: z.string(),
  snippets: z.array(z.string()),
});

export const atsRewriteSuggestionSchema = z.object({
  target: z.string(),
  reason: z.string(),
  before: z.string().nullable().optional(),
  after: z.string(),
});

export const atsSemanticMatchSchema = z.object({
  resumeToJob: z.number().int().min(0).max(100),
  titleAlignment: z.number().int().min(0).max(100),
  requiredSkillCoverage: z.number().int().min(0).max(100),
  embeddingModel: z.string(),
  matchedConcepts: z.array(z.string()),
});

export const atsModelSignalSchema = z.object({
  label: z.string(),
  impact: z.number().int().min(-100).max(100),
  direction: z.enum(["positive", "negative"]),
});

export const atsCompatibilityPredictionSchema = z.object({
  modelVersion: z.string(),
  modelType: z.string(),
  probability: z.number().int().min(0).max(100),
  confidence: z.enum(["low", "medium", "high"]),
  signals: z.array(atsModelSignalSchema),
});

export const atsSkillGapSchema = z.object({
  skill: z.string(),
  severity: z.enum(["critical", "important", "optional"]),
  currentEvidence: z.enum(["missing", "weak", "semantic"]),
  recommendation: z.string(),
  learningFocus: z.string(),
});

export const atsRetrievalCitationSchema = z.object({
  id: z.string(),
  title: z.string(),
  sourceType: z.enum([
    "recruiter_guideline",
    "resume_pattern",
    "domain_knowledge",
  ]),
  relevance: z.number().int().min(0).max(100),
});

export const atsGroundedSuggestionSchema = z.object({
  title: z.string(),
  detail: z.string(),
  citations: z.array(atsRetrievalCitationSchema),
});

const defaultAtsIntelligence = {
  semanticMatch: {
    resumeToJob: 0,
    titleAlignment: 0,
    requiredSkillCoverage: 0,
    embeddingModel: "legacy-analysis",
    matchedConcepts: [],
  },
  compatibilityPrediction: {
    modelVersion: "legacy-analysis",
    modelType: "legacy-analysis",
    probability: 0,
    confidence: "low" as const,
    signals: [],
  },
  skillGaps: [],
  recruiterGuidance: [],
};

export const atsIntelligenceSchema = z
  .object({
    semanticMatch: atsSemanticMatchSchema,
    compatibilityPrediction: atsCompatibilityPredictionSchema,
    skillGaps: z.array(atsSkillGapSchema),
    recruiterGuidance: z.array(atsGroundedSuggestionSchema),
  })
  .default(defaultAtsIntelligence);

export const atsAnalysisResponseSchema = z.object({
  analysisId: z.string().uuid().optional(),
  resumeId: z.string().nullable().optional(),
  overallScore: z.number().int().min(0).max(100),
  verdict: z.enum(["strong_match", "good_match", "partial_match", "needs_work"]),
  categoryScores: atsScoreBreakdownSchema,
  jobProfile: atsJobProfileSchema,
  evidence: z.array(atsKeywordEvidenceSchema),
  matchedKeywords: z.array(z.string()),
  missingKeywords: z.array(z.string()),
  strengths: z.array(z.string()),
  risks: z.array(z.string()),
  suggestions: z.array(atsSuggestionSchema),
  rewriteSuggestions: z.array(atsRewriteSuggestionSchema),
  intelligence: atsIntelligenceSchema,
  summary: z.string(),
});

export type AtsAnalyzeInput = z.infer<typeof atsAnalyzeSchema>;
export type AtsAnalysisResponse = z.infer<typeof atsAnalysisResponseSchema>;

File: apps/ai-service/app/schemas/ats.py
from typing import Literal

from pydantic import BaseModel, Field, model_validator

from .resume import StructuredResume


class AtsAnalyzeRequest(BaseModel):
    geminiApiKey: str | None = None
    resumeId: str | None = None
    fileUrl: str | None = None
    fileName: str | None = None
    resumeText: str | None = Field(default=None, max_length=100_000)
    structuredResume: StructuredResume | None = None
    jobTitle: str = Field(default="", max_length=200)
    companyName: str = Field(default="", max_length=200)
    jobDescription: str = Field(min_length=20, max_length=100_000)
    targetKeywords: list[str] = Field(default_factory=list, max_length=80)

    @model_validator(mode="after")
    def require_resume_source(self):
        has_resume_text = bool(self.resumeText and self.resumeText.strip())
        has_structured_resume = self.structuredResume is not None
        has_file = bool(self.fileUrl)

        if not (has_resume_text or has_structured_resume or has_file):
            raise ValueError(
                "Provide resumeText, structuredResume, or fileUrl."
            )

        self.targetKeywords = [
            keyword.strip()
            for keyword in self.targetKeywords
            if keyword and keyword.strip()
        ]

        return self


class AtsScoreBreakdown(BaseModel):
    keywords: int = Field(ge=0, le=100)
    semantic: int = Field(ge=0, le=100)
    skills: int = Field(ge=0, le=100)
    experience: int = Field(ge=0, le=100)
    impact: int = Field(ge=0, le=100)
    formatting: int = Field(ge=0, le=100)


class AtsSuggestion(BaseModel):
    priority: Literal["high", "medium", "low"]
    title: str
    detail: str
    example: str | None = None


class AtsJobProfile(BaseModel):
    title: str
    seniority: str
    yearsRequired: int
    requiredKeywords: list[str]
    preferredKeywords: list[str]
    responsibilities: list[str]
    certifications: list[str]


class AtsKeywordEvidence(BaseModel):
    keyword: str
    status: Literal["exact", "semantic", "missing"]
    strength: int = Field(ge=0, le=100)
    sourceSection: str
    snippets: list[str]


class AtsRewriteSuggestion(BaseModel):
    target: str
    reason: str
    before: str | None = None
    after: str


class AtsSemanticMatch(BaseModel):
    resumeToJob: int = Field(ge=0, le=100)
    titleAlignment: int = Field(ge=0, le=100)
    requiredSkillCoverage: int = Field(ge=0, le=100)
    embeddingModel: str
    matchedConcepts: list[str]


class AtsModelSignal(BaseModel):
    label: str
    impact: int = Field(ge=-100, le=100)
    direction: Literal["positive", "negative"]


class AtsCompatibilityPrediction(BaseModel):
    modelVersion: str
    modelType: str
    probability: int = Field(ge=0, le=100)
    confidence: Literal["low", "medium", "high"]
    signals: list[AtsModelSignal]


class AtsSkillGap(BaseModel):
    skill: str
    severity: Literal["critical", "important", "optional"]
    currentEvidence: Literal["missing", "weak", "semantic"]
    recommendation: str
    learningFocus: str


class AtsRetrievalCitation(BaseModel):
    id: str
    title: str
    sourceType: Literal["recruiter_guideline", "resume_pattern", "domain_knowledge"]
    relevance: int = Field(ge=0, le=100)


class AtsGroundedSuggestion(BaseModel):
    title: str
    detail: str
    citations: list[AtsRetrievalCitation]


class AtsIntelligence(BaseModel):
    semanticMatch: AtsSemanticMatch
    compatibilityPrediction: AtsCompatibilityPrediction
    skillGaps: list[AtsSkillGap]
    recruiterGuidance: list[AtsGroundedSuggestion]


class AtsAnalysisResponse(BaseModel):
    resumeId: str | None = None
    overallScore: int = Field(ge=0, le=100)
    verdict: Literal["strong_match", "good_match", "partial_match", "needs_work"]
    categoryScores: AtsScoreBreakdown
    jobProfile: AtsJobProfile
    evidence: list[AtsKeywordEvidence]
    matchedKeywords: list[str]
    missingKeywords: list[str]
    strengths: list[str]
    risks: list[str]
    suggestions: list[AtsSuggestion]
    rewriteSuggestions: list[AtsRewriteSuggestion]
    intelligence: AtsIntelligence
    summary: str
